Tests/SourceKitLSPTests/MainFilesProviderTests.swift

```diff
@@ -19,78 +19,197 @@ import SourceKitLSP
-    XCTAssertEqual(ws.index.mainFilesContainingFile(shared_h), [])
-    XCTAssertEqual(ws.index.mainFilesContainingFile(bridging), [d])
+    try await SwiftPMTestWorkspace.build(at: ws.scratchDirectory)
 
-    try await fulfillmentOfOrThrow([mainFilesDelegate.expectation])
+    // After building we know that 'shared.h' is included from 'MyLibrary.c' and thus we use its build settings.
+    let postBuildDiags = try await ws.testClient.nextDiagnosticsNotification()
+    XCTAssertEqual(postBuildDiags.diagnostics.count, 1)
+    let diag = try XCTUnwrap(postBuildDiags.diagnostics.first)
+    XCTAssertEqual(diag.message, "Unused variable 'fromMyLibrary'")
+  }
 
-    XCTAssertEqual(ws.index.mainFilesContainingFile(DocumentURI(string: "not:file")), [])
+  func testMainFileForSharedHeaderOutsideOfTarget() async throws {
+    let ws = try await SwiftPMTestWorkspace(
+      files: [
+        "Sources/shared.h": """
+        void bridging(void) {
+          int VARIABLE_NAME = 1;
+        }
+        """,
+        "Sources/MyLibrary/include/MyLibrary.h": "",
+        "Sources/MyLibrary/MyLibrary.c": """
+        #include "$TEST_DIR/Sources/shared.h"
+        """,
+        "Sources/MyFancyLibrary/include/MyFancyLibrary.h": "",
+        "Sources/MyFancyLibrary/MyFancyLibrary.c": """
+        #include "$TEST_DIR/Sources/shared.h"
+        """,
+      ],
+      manifest: """
+        // swift-tools-version: 5.7
+
+        import PackageDescription
+
+        let package = Package(
+          name: "MyLibrary",
+          targets: [
+            .target(
+              name: "MyLibrary",
+              cSettings: [.define("VARIABLE_NAME", to: "fromMyLibrary"), .unsafeFlags(["-Wunused-variable"])]
+            ),
+            .target(
+              name: "MyFancyLibrary",
+              cSettings: [.define("VARIABLE_NAME", to: "fromMyFancyLibrary"), .unsafeFlags(["-Wunused-variable"])]
+            )
+          ]
+        )
+        """,
+      build: true
+    )
+
+    _ = try ws.openDocument("shared.h", language: .c)
+
+    // We could pick build settings from either 'MyLibrary.c' or 'MyFancyLibrary.c'. We currently pick the
+    // lexicographically first to be deterministic, which is 'MyFancyLibrary'.
+    let diags = try await ws.testClient.nextDiagnosticsNotification()
+    XCTAssertEqual(diags.diagnostics.count, 1)
+    let diag = try XCTUnwrap(diags.diagnostics.first)
+    XCTAssertEqual(diag.message, "Unused variable 'fromMyFancyLibrary'")
+  }
+
+  func testMainFileChangesIfIncludeIsAdded() async throws {
+    let ws = try await SwiftPMTestWorkspace(
+      files: [
+        "Sources/shared.h": """
+        void bridging(void) {
+          int VARIABLE_NAME = 1;
+        }
+        """,
+        "Sources/MyLibrary/include/MyLibrary.h": "",
+        "Sources/MyLibrary/MyLibrary.c": """
+        #include "$TEST_DIR/Sources/shared.h"
+        """,
+        "Sources/MyFancyLibrary/include/MyFancyLibraryh": "",
```

Comment:
This file is unused isn't it?

Reply:
SwiftPM refuses to build if a C target does not have an `include` directory. And in the current infrastructure, the only way to create a directory is to put a file in there.

Reply:
I see. That's... okay. Could just add a `dummy` or something though (to make that more clear)

Reply:
Good idea. Renamed to `dummy.h`.